DecisionTrees/main.ipynb

```diff
@@ -0,0 +1,337 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Decision Trees\n",
+    "Decision trees are a fairly easy to understand machine learning model that works by making use of thresholds on the independent variables to achieve classification (binary or multiclass).\n",
+    "\n",
+    "E.g:\n",
+    "two independent variables: x1 and x2\n",
+    "\n",
+    "three classes for y: 0, 1 and 2\n",
+    "\n",
+    "Our Decision Tree model could be something as simple as:\n",
+    "\n",
+    "```\n",
+    "if x1 < 3:\n",
+    "    y = 2\n",
+    "else:\n",
+    "    if x2 < 5:\n",
+    "        y = 0\n",
+    "    else:\n",
+    "        y = 1\n",
+    "```\n",
+    "\n",
+    "The way these values for which the independent variables are compared against, are determined through entropy: the measurement of randomness."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "0.0\n",
+      "1.0\n",
+      "0.9402859586706311\n",
+      "1.584962500721156\n",
+      "0.9999999999999999\n"
+     ]
+    }
+   ],
+   "source": [
+    "from math import log2\n",
+    "\n",
+    "def entropy(values):\n",
+    "    total = sum(values)\n",
+    "    result = 0\n",
+    "    for val in values:\n",
+    "        if val == 0:\n",
+    "            continue\n",
+    "        prob = val/total\n",
+    "        result += prob*log2(prob)\n",
+    "    result *= -1\n",
+    "    return max(0.0,result)\n",
```

Comment:
I guess that result variable is being abused. Ideally result would be assigned once and returned at the end of a function. Please make sure last line of this function is return result